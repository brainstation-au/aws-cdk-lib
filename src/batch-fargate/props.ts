import { z } from 'zod';

export const BatchFargateProps = z.object({
  vpcId: z.string(),
  subnetIds: z.string().array(),
  maxvCpus: z.number().positive(),
  securityGroupIds: z.string().array(),
});

/* eslint-disable @typescript-eslint/no-redeclare */
export type BatchFargateProps = z.infer<typeof BatchFargateProps>;
