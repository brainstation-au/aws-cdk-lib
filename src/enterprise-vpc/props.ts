import { z } from 'zod';

const Rfc1918 = z.enum([
  '10.0.0.0/8',
  '172.16.0.0/12',
  '192.168.0.0/16',
]);

export const SimpleVPCProps = z.object({
  name: z.string().min(2).max(20),
  ipv4IpamPoolId: z.string().regex(/^ipam-pool-[a-z0-9]{17}$/g),
  ipv4NetmaskLength: z.number().min(16).max(25).default(19),
});

const transit = z.object({
  transitGatewayId: z.string().regex(/^tgw-[a-z0-9]{17}$/g),
  enterpriseCidr: Rfc1918.default(Rfc1918.enum['10.0.0.0/8']),
});

const nat = z.object({
  attachNatGateways: z.literal(true).refine((v) => !!v),
});

export const VPCWithNATProps = SimpleVPCProps.merge(nat);

export const VPCWithTransitProps = SimpleVPCProps.merge(transit);

export const VPCWithTransitAndNATProps = SimpleVPCProps.merge(transit).merge(nat);

export const VPCWithEnterpriseNATProps = SimpleVPCProps.merge(transit).merge(nat).extend({
  transitGatewayRouteTableId: z.string().regex(/^tgw-rtb-[a-z0-9]{17}$/g),
});

export const EnterpriseVPCProps = z.union([
  VPCWithEnterpriseNATProps,
  VPCWithTransitAndNATProps,
  VPCWithTransitProps,
  VPCWithNATProps,
  SimpleVPCProps,
]);

/* eslint-disable @typescript-eslint/no-redeclare */
export type SimpleVPCProps = z.infer<typeof SimpleVPCProps>;
export type VPCWithNATProps = z.infer<typeof VPCWithNATProps>;
export type VPCWithTransitProps = z.infer<typeof VPCWithTransitProps>;
export type VPCWithTransitAndNATProps = z.infer<typeof VPCWithTransitAndNATProps>;
export type VPCWithEnterpriseNATProps = z.infer<typeof VPCWithEnterpriseNATProps>;
export type EnterpriseVPCProps = z.infer<typeof EnterpriseVPCProps>;
