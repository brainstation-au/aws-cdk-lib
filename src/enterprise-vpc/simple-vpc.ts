import { Construct } from 'constructs';
import { EnterpriseVPC } from './enterprise-vpc';
import { SimpleVPCProps } from './props';

export class SimpleVPC extends EnterpriseVPC {
  constructor(scope: Construct, id: string, props: SimpleVPCProps) {
    super(scope, id, SimpleVPCProps.parse(props));
  }
}
