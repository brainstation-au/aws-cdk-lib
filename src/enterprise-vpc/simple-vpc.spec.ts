import * as cdk from 'aws-cdk-lib';
import YAML from 'yaml';
import { SimpleVPC } from './simple-vpc';
import { SimpleVPCProps } from './props';

const createStack = (vpcProps: SimpleVPCProps) => {
  class MyStack extends cdk.Stack {
    constructor(scope: cdk.App, id: string, props?:cdk.StackProps) {
      super(scope, id, props);

      const name = new cdk.CfnParameter(this, 'GroupName', {
        type: 'String',
        default: vpcProps.name,
      });

      new SimpleVPC(this, 'MyVpc', {
        ...vpcProps,
        name: cdk.Fn.ref(name.logicalId),
      });
    }
  }

  const app = new cdk.App();
  return new MyStack(app, 'MyStack');
};

describe('simple-vpc', () => {
  const stack = createStack({
    name: 'brainstation',
    ipv4IpamPoolId: 'ipam-pool-0fdbd5d6bea6196ed',
    ipv4NetmaskLength: 19,
  });
  const template = cdk.assertions.Template.fromStack(stack);

  test('produces correct cloudformation template', () => {
    expect(YAML.stringify(template.toJSON())).toMatchSnapshot();
  });

  test('creates a VPC', () => {
    template.resourceCountIs('AWS::EC2::VPC', 1);
  });

  test('creates an internet gateway', () => {
    template.resourceCountIs('AWS::EC2::InternetGateway', 1);
  });

  test('creates an internet gateway attachment', () => {
    template.resourceCountIs('AWS::EC2::VPCGatewayAttachment', 1);
  });

  test('creates 6 subnets', () => {
    template.resourceCountIs('AWS::EC2::Subnet', 6);
  });

  test('does not create any elastic IP', () => {
    template.resourceCountIs('AWS::EC2::EIP', 0);
  });

  test('does not create any NAT Gateway', () => {
    template.resourceCountIs('AWS::EC2::NatGateway', 0);
  });

  test('does not create any transit gateway attachment', () => {
    template.resourceCountIs('AWS::EC2::TransitGatewayAttachment', 0);
  });

  test('creates 2 route table', () => {
    template.resourceCountIs('AWS::EC2::RouteTable', 2);
  });

  test('creates 6 subnet route table association', () => {
    template.resourceCountIs('AWS::EC2::SubnetRouteTableAssociation', 6);
  });

  test('creates 1 routes', () => {
    template.resourceCountIs('AWS::EC2::Route', 1);
  });
});
