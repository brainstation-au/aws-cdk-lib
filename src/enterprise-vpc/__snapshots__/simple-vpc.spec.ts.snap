// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple-vpc produces correct cloudformation template 1`] = `
"Parameters:
  GroupName:
    Type: String
    Default: brainstation
  BootstrapVersion:
    Type: AWS::SSM::Parameter::Value<String>
    Default: /cdk-bootstrap/hnb659fds/version
    Description: Version of the CDK Bootstrap resources in this environment,
      automatically retrieved from SSM Parameter Store. [cdk:skip]
Resources:
  MyVpcVPC4AF5F27E:
    Type: AWS::EC2::VPC
    Properties:
      EnableDnsHostnames: true
      EnableDnsSupport: true
      Ipv4IpamPoolId: ipam-pool-0fdbd5d6bea6196ed
      Ipv4NetmaskLength: 19
      Tags:
        - Key: Name
          Value:
            Ref: GroupName
  MyVpcInternetGateway475020C9:
    Type: AWS::EC2::InternetGateway
    Properties:
      Tags:
        - Key: Name
          Value:
            Ref: GroupName
  MyVpcInternetGatewayAttachmentF5FABB76:
    Type: AWS::EC2::VPCGatewayAttachment
    Properties:
      VpcId:
        Ref: MyVpcVPC4AF5F27E
      InternetGatewayId:
        Ref: MyVpcInternetGateway475020C9
  MyVpcSubnetPublicAFFDB0540:
    Type: AWS::EC2::Subnet
    Properties:
      VpcId:
        Ref: MyVpcVPC4AF5F27E
      AvailabilityZone:
        Fn::Select:
          - 0
          - Fn::GetAZs: ""
      CidrBlock:
        Fn::Select:
          - 0
          - Fn::Cidr:
              - Fn::GetAtt:
                  - MyVpcVPC4AF5F27E
                  - CidrBlock
              - 8
              - "10"
      MapPublicIpOnLaunch: true
      Tags:
        - Key: Name
          Value:
            Fn::Join:
              - ""
              - - Ref: GroupName
                - /public-a
  MyVpcSubnetPublicB638F8D5A:
    Type: AWS::EC2::Subnet
    Properties:
      VpcId:
        Ref: MyVpcVPC4AF5F27E
      AvailabilityZone:
        Fn::Select:
          - 1
          - Fn::GetAZs: ""
      CidrBlock:
        Fn::Select:
          - 1
          - Fn::Cidr:
              - Fn::GetAtt:
                  - MyVpcVPC4AF5F27E
                  - CidrBlock
              - 8
              - "10"
      MapPublicIpOnLaunch: true
      Tags:
        - Key: Name
          Value:
            Fn::Join:
              - ""
              - - Ref: GroupName
                - /public-b
  MyVpcSubnetPublicCB7352644:
    Type: AWS::EC2::Subnet
    Properties:
      VpcId:
        Ref: MyVpcVPC4AF5F27E
      AvailabilityZone:
        Fn::Select:
          - 2
          - Fn::GetAZs: ""
      CidrBlock:
        Fn::Select:
          - 2
          - Fn::Cidr:
              - Fn::GetAtt:
                  - MyVpcVPC4AF5F27E
                  - CidrBlock
              - 8
              - "10"
      MapPublicIpOnLaunch: true
      Tags:
        - Key: Name
          Value:
            Fn::Join:
              - ""
              - - Ref: GroupName
                - /public-c
  MyVpcSubnetPrivateA0159B61C:
    Type: AWS::EC2::Subnet
    Properties:
      VpcId:
        Ref: MyVpcVPC4AF5F27E
      AvailabilityZone:
        Fn::Select:
          - 0
          - Fn::GetAZs: ""
      CidrBlock:
        Fn::Select:
          - 3
          - Fn::Cidr:
              - Fn::GetAtt:
                  - MyVpcVPC4AF5F27E
                  - CidrBlock
              - 8
              - "10"
      MapPublicIpOnLaunch: false
      Tags:
        - Key: Name
          Value:
            Fn::Join:
              - ""
              - - Ref: GroupName
                - /private-a
  MyVpcSubnetPrivateB19604239:
    Type: AWS::EC2::Subnet
    Properties:
      VpcId:
        Ref: MyVpcVPC4AF5F27E
      AvailabilityZone:
        Fn::Select:
          - 1
          - Fn::GetAZs: ""
      CidrBlock:
        Fn::Select:
          - 4
          - Fn::Cidr:
              - Fn::GetAtt:
                  - MyVpcVPC4AF5F27E
                  - CidrBlock
              - 8
              - "10"
      MapPublicIpOnLaunch: false
      Tags:
        - Key: Name
          Value:
            Fn::Join:
              - ""
              - - Ref: GroupName
                - /private-b
  MyVpcSubnetPrivateC914D6E3A:
    Type: AWS::EC2::Subnet
    Properties:
      VpcId:
        Ref: MyVpcVPC4AF5F27E
      AvailabilityZone:
        Fn::Select:
          - 2
          - Fn::GetAZs: ""
      CidrBlock:
        Fn::Select:
          - 5
          - Fn::Cidr:
              - Fn::GetAtt:
                  - MyVpcVPC4AF5F27E
                  - CidrBlock
              - 8
              - "10"
      MapPublicIpOnLaunch: false
      Tags:
        - Key: Name
          Value:
            Fn::Join:
              - ""
              - - Ref: GroupName
                - /private-c
  MyVpcRouteTablePublic20E2D0AF:
    Type: AWS::EC2::RouteTable
    Properties:
      VpcId:
        Ref: MyVpcVPC4AF5F27E
      Tags:
        - Key: Name
          Value:
            Fn::Join:
              - ""
              - - Ref: GroupName
                - /public
  MyVpcRoutePublicDefault4074E1DC:
    Type: AWS::EC2::Route
    Properties:
      RouteTableId:
        Ref: MyVpcRouteTablePublic20E2D0AF
      DestinationCidrBlock: 0.0.0.0/0
      GatewayId:
        Ref: MyVpcInternetGateway475020C9
    DependsOn:
      - MyVpcInternetGatewayAttachmentF5FABB76
  MyVpcSubnetRouteTableAssociationPublicAFF4F72C0:
    Type: AWS::EC2::SubnetRouteTableAssociation
    Properties:
      RouteTableId:
        Ref: MyVpcRouteTablePublic20E2D0AF
      SubnetId:
        Ref: MyVpcSubnetPublicAFFDB0540
  MyVpcSubnetRouteTableAssociationPublicB50387D00:
    Type: AWS::EC2::SubnetRouteTableAssociation
    Properties:
      RouteTableId:
        Ref: MyVpcRouteTablePublic20E2D0AF
      SubnetId:
        Ref: MyVpcSubnetPublicB638F8D5A
  MyVpcSubnetRouteTableAssociationPublicC14D4C18D:
    Type: AWS::EC2::SubnetRouteTableAssociation
    Properties:
      RouteTableId:
        Ref: MyVpcRouteTablePublic20E2D0AF
      SubnetId:
        Ref: MyVpcSubnetPublicCB7352644
  MyVpcRouteTablePrivate96A3F999:
    Type: AWS::EC2::RouteTable
    Properties:
      VpcId:
        Ref: MyVpcVPC4AF5F27E
      Tags:
        - Key: Name
          Value:
            Fn::Join:
              - ""
              - - Ref: GroupName
                - /private
  MyVpcSubnetRouteTableAssociationPrivateAEC013698:
    Type: AWS::EC2::SubnetRouteTableAssociation
    Properties:
      RouteTableId:
        Ref: MyVpcRouteTablePrivate96A3F999
      SubnetId:
        Ref: MyVpcSubnetPrivateA0159B61C
  MyVpcSubnetRouteTableAssociationPrivateB02AA7F96:
    Type: AWS::EC2::SubnetRouteTableAssociation
    Properties:
      RouteTableId:
        Ref: MyVpcRouteTablePrivate96A3F999
      SubnetId:
        Ref: MyVpcSubnetPrivateB19604239
  MyVpcSubnetRouteTableAssociationPrivateC3AA5A2D3:
    Type: AWS::EC2::SubnetRouteTableAssociation
    Properties:
      RouteTableId:
        Ref: MyVpcRouteTablePrivate96A3F999
      SubnetId:
        Ref: MyVpcSubnetPrivateC914D6E3A
  MyVpcParamVpcIdCCE5D477:
    Type: AWS::SSM::Parameter
    Properties:
      Type: String
      Value:
        Ref: MyVpcVPC4AF5F27E
      Description: VPC Id
      Name:
        Fn::Join:
          - ""
          - - /cf/
            - Ref: GroupName
            - /vpc/id
  MyVpcParamSecurityGroupId898A4065:
    Type: AWS::SSM::Parameter
    Properties:
      Type: String
      Value:
        Fn::GetAtt:
          - MyVpcVPC4AF5F27E
          - DefaultSecurityGroup
      Description: ID of default securite group for intra-vpc access
      Name:
        Fn::Join:
          - ""
          - - /cf/
            - Ref: GroupName
            - /security-group/default/id
  MyVpcParamPublicSubnetsIds71498C92:
    Type: AWS::SSM::Parameter
    Properties:
      Type: List<String>
      Value:
        Fn::Join:
          - ","
          - - Ref: MyVpcSubnetPublicAFFDB0540
            - Ref: MyVpcSubnetPublicB638F8D5A
            - Ref: MyVpcSubnetPublicCB7352644
      Description: List of IDs of the public subnets
      Name:
        Fn::Join:
          - ""
          - - /cf/
            - Ref: GroupName
            - /public-subnets/ids
  MyVpcParamPrivateSubnetsIds762F1123:
    Type: AWS::SSM::Parameter
    Properties:
      Type: List<String>
      Value:
        Fn::Join:
          - ","
          - - Ref: MyVpcSubnetPrivateA0159B61C
            - Ref: MyVpcSubnetPrivateB19604239
            - Ref: MyVpcSubnetPrivateC914D6E3A
      Description: List of IDs of the private subnets
      Name:
        Fn::Join:
          - ""
          - - /cf/
            - Ref: GroupName
            - /private-subnets/ids
Outputs:
  MyVpcVpcIdEA080D3B:
    Description: VPC ID
    Value:
      Ref: MyVpcVPC4AF5F27E
    Export:
      Name:
        Fn::Join:
          - ""
          - - Ref: GroupName
            - -vpc
  MyVpcPublicSubnets7A049237:
    Description: List of IDs of the public subnets
    Value:
      Fn::Join:
        - ","
        - - Ref: MyVpcSubnetPublicAFFDB0540
          - Ref: MyVpcSubnetPublicB638F8D5A
          - Ref: MyVpcSubnetPublicCB7352644
    Export:
      Name:
        Fn::Join:
          - ""
          - - Ref: GroupName
            - -public-subnets
  MyVpcPrivateSubnets223FF8A8:
    Description: List of IDs of the private subnets
    Value:
      Fn::Join:
        - ","
        - - Ref: MyVpcSubnetPrivateA0159B61C
          - Ref: MyVpcSubnetPrivateB19604239
          - Ref: MyVpcSubnetPrivateC914D6E3A
    Export:
      Name:
        Fn::Join:
          - ""
          - - Ref: GroupName
            - -private-subnets
Rules:
  CheckBootstrapVersion:
    Assertions:
      - Assert:
          Fn::Not:
            - Fn::Contains:
                - - "1"
                  - "2"
                  - "3"
                  - "4"
                  - "5"
                - Ref: BootstrapVersion
        AssertDescription: CDK bootstrap stack version 6 required. Please run 'cdk
          bootstrap' with a recent version of the CDK CLI.
"
`;
